File: 621.java
// 621. Task Scheduler
// Given a characters array tasks, representing the tasks a CPU needs to do, where each letter represents a different task. Tasks could be done in any order. Each task is done in one unit of time. For each unit of time, the CPU could complete either one task or just be idle.
//
// However, there is a non-negative integer n that represents the cooldown period between two same tasks (the same letter in the array), that is that there must be at least n units of time between any two same tasks.
//
// Return the least number of units of times that the CPU will take to finish all the given tasks.
//
//
//
// Example 1:
//
// Input: tasks = ["A","A","A","B","B","B"], n = 2
// Output: 8
// Explanation:
// A -> B -> idle -> A -> B -> idle -> A -> B
// There is at least 2 units of time between any two same tasks.
// Example 2:
//
// Input: tasks = ["A","A","A","B","B","B"], n = 0
// Output: 6
// Explanation: On this case any permutation of size 6 would work since n = 0.
// ["A","A","A","B","B","B"]
// ["A","B","A","B","A","B"]
// ["B","B","B","A","A","A"]
// ...
// And so on.
// Example 3:
//
// Input: tasks = ["A","A","A","A","A","A","B","C","D","E","F","G"], n = 2
// Output: 16
// Explanation:
// One possible solution is
// A -> B -> C -> A -> D -> E -> A -> F -> G -> A -> idle -> idle -> A -> idle -> idle -> A
//
//
// Constraints:
//
// 1 <= task.length <= 104
// tasks[i] is upper-case English letter.
// The integer n is in the range [0, 100].
//
// Runtime: 3 ms, faster than 61.13% of Java online submissions for Task Scheduler.
// Memory Usage: 40.7 MB, less than 23.02% of Java online submissions for Task Scheduler.
class Solution {
    public int leastInterval(char[] tasks, int n) {
        if (n == 0) return tasks.length;
        int max = 0;
        int maxCounter = 0;
        char[] taskArr = new char[26];
        for (char task: tasks) {
            taskArr[task - 'A']++;
            if (taskArr[task - 'A'] == max) maxCounter++;
            else if (taskArr[task - 'A'] > max) {
                max = taskArr[task - 'A'];
                maxCounter = 1;
            }
        }
        return tasks.length + Math.max((max - 1) * (n - maxCounter + 1) - tasks.length + maxCounter * max, 0);
    }
}

// n = 0 -> time = task.length
// ["A","A","A","B","B","B"] n =2
// A: 3
// B: 3
// A x x A x x A 
// A B x A B x A B

// length = tasks.length + idles
//        = tasks.length + (slots - non max occurance tasks or 0)
// max occurance * (n + 1)

File: go/621.go
// 621. Task Scheduler

// You are given an array of CPU tasks, each labeled with a letter from A to Z, and a number n. Each CPU interval can be idle or allow the completion of one task. Tasks can be completed in any order, but there's a constraint: there has to be a gap of at least n intervals between two tasks with the same label.

// Return the minimum number of CPU intervals required to complete all tasks.

 

// Example 1:

// Input: tasks = ["A","A","A","B","B","B"], n = 2

// Output: 8

// Explanation: A possible sequence is: A -> B -> idle -> A -> B -> idle -> A -> B.

// After completing task A, you must wait two intervals before doing A again. The same applies to task B. In the 3rd interval, neither A nor B can be done, so you idle. By the 4th interval, you can do A again as 2 intervals have passed.

// Example 2:

// Input: tasks = ["A","C","A","B","D","B"], n = 1

// Output: 6

// Explanation: A possible sequence is: A -> B -> C -> D -> A -> B.

// With a cooling interval of 1, you can repeat a task after just one other task.

// Example 3:

// Input: tasks = ["A","A","A", "B","B","B"], n = 3

// Output: 10

// Explanation: A possible sequence is: A -> B -> idle -> idle -> A -> B -> idle -> idle -> A -> B.

// There are only two types of tasks, A and B, which need to be separated by 3 intervals. This leads to idling twice between repetitions of these tasks.

 

// Constraints:

// 1 <= tasks.length <= 104
// tasks[i] is an uppercase English letter.
// 0 <= n <= 100

// Runtime 2 ms Beats 76.96%
// Memory 8.06 MB Beats 84.29%
func leastInterval(tasks []byte, n int) int {
    maxCount := 0
    countOfMax := 0
    counts := make([]int, 26)
    for i := range tasks {
        index := tasks[i] - 'A'
        counts[index]++
        if counts[index] == maxCount {
            countOfMax++
        } else if maxCount < counts[index] {
            maxCount = counts[index]
            countOfMax = 1
        }
    }
    idles := (maxCount - 1) * (n - countOfMax + 1) - (len(tasks) - maxCount * countOfMax)
    if idles < 0 {
        idles = 0
    }
    return len(tasks) + idles
}